contracts/ibc-callbacks/src/contract.rs

```diff
@@ -0,0 +1,174 @@
+    src_callback: Option<IbcSrcCallback>,
+    dst_callback: Option<IbcDstCallback>,
+}
+
+impl IbcCallbackData {
+    /// Use this if you want to execute callbacks on both the source and destination chain.
+    ///
+    /// In the first prototype, we only support receiving callbacks on the source chain.
+    pub fn both(src_callback: IbcSrcCallback, dst_callback: IbcDstCallback) -> Self {
+        IbcCallbackData {
+            src_callback: Some(src_callback),
+            dst_callback: Some(dst_callback),
+        }
+    }
+
+    /// Use this if you want to execute callbacks on the source chain, but not the destination chain.
+    pub fn source(src_callback: IbcSrcCallback) -> Self {
+        IbcCallbackData {
+            src_callback: Some(src_callback),
+            dst_callback: None,
+        }
+    }
+
+    /// Use this if you want to execute callbacks on the destination chain, but not the source chain.
+    ///
+    /// In the first prototype, we only support receiving callbacks on the source chain.
+    pub fn destination(dst_callback: IbcDstCallback) -> Self {
+        IbcCallbackData {
+            src_callback: None,
+            dst_callback: Some(dst_callback),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct IbcSrcCallback {
+    /// The source chain address that should receive the callback.
+    /// You probably want to put `env.contract.address` here.
+    pub address: Addr,
+    /// Optional gas limit for the callback (in Cosmos SDK gas units)
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub gas_limit: Option<Uint64>,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct IbcDstCallback {
+    /// The destination chain address that should receive the callback.
+    pub address: String,
+    /// Optional gas limit for the callback (in Cosmos SDK gas units)
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub gas_limit: Option<Uint64>,
+}
+
+/// The type of IBC callback that is being called.
+///
+/// IBC callbacks are needed for cases where your contract triggers the sending of an IBC packet
+/// through some other message (i.e. not through [`IbcMsg::SendPacket`]) and needs to know whether
+/// or not the packet was successfully received on the other chain.
+/// A prominent example is the [`IbcMsg::Transfer`] message. Without callbacks, you cannot know
+/// whether the transfer was successful or not.
+///
+/// Note that there are some prerequisites that need to be fulfilled to receive source chain callbacks:
+/// - The contract must implement the `ibc_source_chain_callback` entrypoint.
+/// - The module that sends the packet must be wrapped by an `IBCMiddleware`
+///   (i.e. the source chain needs to support callbacks for the message you are sending).
+/// - You have to add json-encoded [`IbcCallbackData`] to a specific field of the message.
+///   For `IbcMsg::Transfer`, this is the `memo` field.
```

Comment:
🥲 But why? Why does the other chain even need to know I want this callback?

Reply:
The callbacks middleware on the source chain checks for this to decide whether to call the source chain callbacks (and I guess also on the destination chain for the ReceivePacketCallback, but I don't implement that for the first prototype, as Serdar mentioned it's a bit tricky to implement securely).